migration/migration.go

```diff
@@ -0,0 +1,175 @@
+package migration
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+)
+
+var (
+	log = logger.Logger("migrations")
+
+	keyCurrentVersion = datastore.NewKey("version")
+)
+
+// Migrator ensures a datastore goes through all the needed
+// Migrations to upgrade its current version to the latest version.
+type Migrator struct {
+	ds         datastore.TxnDatastore
+	migrations map[int]Migration
+}
+
+// Migration runs a vA->v(A+1) migration.
+type Migration func(datastore.Txn) error
```

Comment:
A `Migration` is any function that receives a `txn` and does whatever it wants to migrate from version X to (X+1).
This function doesn't have to commit the txn, just do its logic. The library will automatically commit or discard the transaction depending if the migration returned an error or not.